lib/Sema/TypeCheckDecl.cpp

```diff
@@ -5597,6 +5600,153 @@ static Optional<ObjCReason> shouldMarkClassAsObjC(TypeChecker &TC,
   return None;
 }
 
+bool checkNominalDeclCircularity(NominalTypeDecl *nominal,
+    llvm::SmallPtrSet<NominalTypeDecl *, 16> &known);
+
+
+// A tuple type have elements with distinct types. A bound generic type has
+// generic arguments. These associated types may form circularities with their
+// declaring context. This function look for nominal types among them.
+bool deconstructTypeForDeclCircularity(Type type,
+    llvm::SmallPtrSet<NominalTypeDecl *, 16> &known) {
+
+  // dive into nominal types for potential circularities.
+  if (auto N = type->getAnyNominal()) {
+    if (checkNominalDeclCircularity(N, known)) {
+      return true;
+    }
+  }
+
+  // check each generic argument types
+  if (auto BGT = type->getAs<BoundGenericType>()) {
+    for (auto T: BGT->getGenericArgs()) {
+      if (deconstructTypeForDeclCircularity(T, known)) {
+        return true;
+      }
+    }
+  }
+
+  // if this is a tuple, deconstruct it and check each element's type
+  if (auto tuple = type->getAs<TupleType>()) {
+    for (auto Elt: tuple->getElements()) {
+      if (deconstructTypeForDeclCircularity(Elt.getType(), known)) {
+        return true;
+      }
+    }
+  }
+
+  return false;
+}
+
+bool checkVarDeclCircularity(VarDecl *vd,
+    llvm::SmallPtrSet<NominalTypeDecl *, 16> &known) {
+  assert(isa<VarDecl>(vd) && "Expected VarDecl");
+  if (vd->isStatic() || !vd->hasStorage() || !vd->hasType()) {
+    return false;
+  }
+  auto vdt = vd->getType();
+  if (vdt->hasReferenceSemantics()) {
+    // reference types are fine
+    return false;
+  }
+  if (auto vdtn = vdt->getAnyNominal()) {
+    if (known.count(vdtn)) {
+      return true;
+    }
+  }
+  if (deconstructTypeForDeclCircularity(vdt, known)) {
+    return true;
+  }
+  return false;
+}
+
+bool checkEnumEltDeclCircularity(EnumElementDecl *eed,
+    llvm::SmallPtrSet<NominalTypeDecl *, 16> &known) {
+  assert(isa<EnumElementDecl>(eed) && "Expected EnumElementDecl");
+  auto eedat = eed->getArgumentType();
+  // indirect cases, cases without arugument types are fine, skip these
+  if (eed->isIndirect() || !eedat) {
+    return false;
+  }
+
+  if (auto eedatn = eedat->getAnyNominal()) {
```

Comment:
Is this necessary given that you call deconstructTypeForDeclCircularity() below? It will catch cases like ```enum E { case X(E) }``` but not ```enum E { case X(E, E) }```